capellambse_context_diagrams/collectors/default.py

```diff
@@ -259,3 +267,91 @@ def port_context_collector(
             info.ports.append(port)
 
     return iter(ctx.values())
+
+
+def add_derived_components_and_interfaces(
+    diagram: context.ContextDiagram, data: _elkjs.ELKInputData
+) -> None:
+    """Add hidden Boxes and Exchanges to ``obj``'s context.
+
+    The derived exchanges are displayed with a dashed line.
+    """
+    if not (derivator := DERIVATORS.get(type(diagram.target))):
+        return
+
+    derivator(diagram, data)
+
+
+def derive_from_functions(
+    diagram: context.ContextDiagram, data: _elkjs.ELKInputData
+) -> None:
+    """Derive Components from allocated functions of the context target.
+
+    A Component, a ComponentExchange and two ComponentPorts are added
+    to ``data``. These elements are prefixed with ``Derived-`` to
+    receive special styling in the serialization step.
+    """
+    assert isinstance(diagram.target, cs.Component)
+    ports = []
+    for fnc in diagram.target.allocated_functions:
+        ports.extend(port_collector(fnc, diagram.type))
+
+    context_box_ids = {child["id"] for child in data["children"]}
+    components: dict[str, cs.Component] = {}
+    for port in ports:
+        for fex in port.exchanges:
+            if isinstance(port, fa.FunctionOutputPort):
+                attr = "target"
+            else:
+                attr = "source"
+
+            try:
+                derived_comp = getattr(fex, attr).owner.owner
+                if (
+                    derived_comp == diagram.target
+                    or derived_comp.uuid in context_box_ids
+                ):
+                    continue
+
+                if derived_comp.uuid not in components:
+                    components[derived_comp.uuid] = derived_comp
+            except AttributeError:
+                ...
+
+    # TODO: Even out derived interfaces on each side
+
+    centerbox = data["children"][0]
+    for i, (uuid, derived_component) in enumerate(components.items(), 1):
+        box = makers.make_box(
+            derived_component,
+            no_symbol=diagram.display_symbols_as_boxes,
+        )
+        class_ = type(derived_comp).__name__
+        box["id"] = f"__Derived-{class_}:{uuid}"
+        data["children"].append(box)
+        source_id = f"__Derived-CP_INOUT:{i}"
+        target_id = f"__Derived-CP_INOUT:{-i}"
+        box.setdefault("ports", []).append(makers.make_port(source_id))
+        centerbox.setdefault("ports", []).append(makers.make_port(target_id))
+        if i % 2 == 0:
+            source_id, target_id = target_id, source_id
```

Comment:
What is the reason for swapping the source and target ids here? So it would be a good idea to add a short comment here explaining why the edge direction is swapped in this case. Also, it seems counterintuitive that we first use the source and target ids to set the ports on the box and centerbox, and then swap them, i.e. if we swap the ids, the box will have the target id as the port and the centerbox will have the source id, even though we set it differently.

Reply:
This is a way to trick ELK in routing the edge. ELK Layered is all about flow direction. If we reverse the direction the edge is routed like the other, resulting in shorter edges with less bendpoints. We do it for every even edge, to even it out on every side.